Created LIP: Define mainnet configuration and migration for Lisk Core v4 (#141)
LiskHQ/lips: proposals/mainnet_migration.md

```
LIP: <LIP number>
Title: Define Mainnet Configuration and Migration for Lisk Core v4
Author: Andreas Kendziorra <andreas.kendziorra@lightcurve.io>
Discussions-To: https://research.lisk.com/t/define-mainnet-configuration-and-migration-for-lisk-core-4/340
Type: Standards Track
Created: <YYYY-MM-DD>
Updated: <YYYY-MM-DD>
Requires: 0060
```

## Abstract

This proposal defines the configuration of Lisk Core v4, including the selection of modules and the choice of some configurable constants. Moreover, it defines the migration process from Lisk Core v3 to Lisk Core v4\. As for the previous hard fork, a snapshot block is used that contains a snapshot of the state. This snapshot block is then treated like a genesis block by Lisk Core v4 nodes. In contrast to the previous hard fork, the existing block and transaction history is not discarded.

## Copyright

This LIP is licensed under the [Creative Commons Zero 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/).

## Motivation

> This LIP is motivated by two points. The first is that the Lisk SDK 6 introduces several new configurable settings. This includes the set of existing modules that can or even must be registered. But also some new constants that must be specified for each chain. Thus, these configurations must be specified for Lisk Core v4 as well which shall be done within this LIP.

Author: Fixed

   </td>
   <td>Module ID of the <a href="https://github.com/LiskHQ/lips/blob/main/proposals/lip-0051.md">Token Module</a>
   </td>
  </tr>
  <tr>
   <td><code>MODULE_ID_LEGACY</code>
   </td>
   <td>uint32
   </td>
   <td>TBD
   </td>
   <td>Module ID of the <a href="https://github.com/LiskHQ/lips/blob/main/proposals/lip-0050.md">Legacy Module</a>.
   </td>
  </tr>
  <tr>
   <td><code>INVALID_ED25519_KEY</code>
   </td>
   <td>bytes
   </td>
   <td>

```   
0x7fffffff
ffffffffff
ffffffffff
ffffffffff
ffffffffff
ffffffffff
ffffff
```

   </td>
   <td>An Ed25519 public key for which the signature validation always fails. This value is used for the <code>generatorPublicKey</code> property of validators in the snapshot block for which no public key is contained within the history since the last snapshot block.
   </td>
  </tr>
  <tr>
   <td><code>DPOS_INIT_ROUNDS</code>
   </td>
   <td>uint32
   </td>
   <td>60,480
   </td>
   <td>The number of rounds for the bootstrap period following the snapshot block. This number corresponds to one week assuming no missed blocks.
   </td>
  </tr>
  <tr>
   <td><code>HEIGHT_SNAPSHOT</code>
   </td>
   <td>uint32
   </td>
   <td>TBD
   </td>

>    <td>The height of the block from which a state snapshot is taken. This block must be an end of a round block. The snapshot block has then the height <code>HEIGHT_SNAPSHOT +1.</code>

Reviewer: ```suggestion
   <td>The height of the block from which a state snapshot is taken. This block must be an end of a round block. The snapshot block has then the height `HEIGHT_SNAPSHOT +1`.
```

Reviewer: I think the conversion tool converts tables with html tags rather than pure markdown tables. All good, I resolve it.

    2. Nodes continue processing blocks until the block at height `HEIGHT_SNAPSHOT` is final.
    3. Once the block at height `HEIGHT_SNAPSHOT` is final, nodes can stop forging and processing new blocks. Finalized blocks with a height larger than or equal to `HEIGHT_SNAPSHOT+1` are discarded.
2. Nodes compute a snapshot block as defined [below](#snapshot-block) using a _migrator tool_ and persist this one locally.
3. Nodes start to use Lisk Core v4, where the steps described in the section [Starting Lisk Core v4](#starting-lisk-core-4) are executed.
4. Once the timeslot of the snapshot block at height `HEIGHT_SNAPSHOT+1` is passed, the first round following the new protocol starts.

##### Starting Lisk Core v4

When Lisk Core v4 is started for the first time, the following steps are performed:

1. Get the snapshot block (the one for height `HEIGHT_SNAPSHOT+1`):
    1. Check if the snapshot block for height `HEIGHT_SNAPSHOT+1` exists locally. If yes, fetch this block. If not, stop here.
2. Process the snapshot block as described in [LIP 0060](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0060.md).

> 3. Check if a database from Lisk Core v3 containing all blocks between heights `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` and `HEIGHT_SNAPSHOT` (inclusive) can be found locally. If yes:

Reviewer: ```suggestion
3. Check if all blocks between heights `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` and `HEIGHT_SNAPSHOT` (inclusive) from Lisk Core v3 can be found locally. If yes:
```

    1. Fetch these blocks from highest height to lowest. Each block is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block and its transactions are persisted in the database.
    2. Skip steps 4 and 5.

> 4. Fetch all blocks between heights `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK+1` and`HEIGHT_SNAPSHOT` (inclusive) via peer-to-peer from highest height to lowest. Each block is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block and its transactions are persisted in the database.

Author: Typo fixed.

> 5. The snapshot block for the height `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` is downloaded from a server. The URL for the source can be configured in Lisk Core v4+. When downloaded, it is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block is persisted in the database.

Reviewer: ```suggestion
5. The snapshot block for the height `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK` is downloaded from a server. The URL for the source can be configured. When downloaded, it is validated using [minimal validation steps as defined below](#minimal-validation-of-core-3-blocks). If this validation step passes, the block is persisted in the database.
```

Reviewer: I don't understand why the part " in Lisk Core v4+" is necessary or correct here as it is talking about Lisk Core 4.

Author: The snapshot block can be downloaded from the server only after it is finalized. So here Lisk Core v4+ means a later version of Lisk Core v4, in which a link to the snapshot block is hardcoded.

This is discussed in details a bit later, in the section Fetching Snapshot Blocks, I have added a reference for clarity.

Reviewer: But here we talk about the previous snapshot block (not that one at `HEIGHT_SNAPSHOT+1`, but the one at `HEIGHT_PREVIOUS_SNAPSHOT_BLOCK`). This block is from August last year and already final. In particular, it would be safe to obtain it from an untrusted source due to the minimal validations defined below.

Author: Ah, you are right! Changed as per your suggestion.

Due to step 1.i, it is a requirement to run Lisk Core v3 and the migrator tool before running Lisk Core v4\. However, nodes starting some time after the migration may fetch the snapshot block and its preceding blocks without running Lisk Core v3 and migrator tool before, as described in the following subsection.

###### Fetching Block History and Snapshot Block without Running Lisk Core v3

> Once the snapshot block at height `HEIGHT_SNAPSHOT+1` is final, a new version of Lisk Core v4 can be released that has the block ID of this snapshot block hard-coded. In the following, we denote this version by Lisk Core v4+. When a Lisk Core v4+ starts for the first time, the same steps as described above are executed, except that step 1 is replaced by the following:

Reviewer: ```suggestion
Once the snapshot block at height `HEIGHT_SNAPSHOT+1` is final, a new version of Lisk Core v4 can be released that has the block ID of this snapshot block hard-coded. In the following, we denote this version by Lisk Core v4+. When Lisk Core v4+ starts for the first time, the same steps as described above are executed, except that step 1 is replaced by the following:
```

Author: Replaced all `HEIGHT_SNAPSHOT+1` by `HEIGHT_SNAPSHOT + 1`.

1. Get the snapshot block (the one for height `HEIGHT_SNAPSHOT+1`):
    1. Check if the snapshot block for height `HEIGHT_SNAPSHOT+1` exists locally. If yes:
        1. Fetch this block.
        2. Verify that the block ID of this block is matching with the hard-coded block ID of the snapshot block. If yes, skip step b. If not, continue with step ii.
    2. The snapshot block for height `HEIGHT_SNAPSHOT+1` is downloaded from a server. The URL for the source can be configured in Lisk Core v4+. When downloaded, it must first be verified that the block ID of this block is matching with the hard-coded block ID of the snapshot block. If not, stop here (the process should be repeated, but the node operator should specify a new server to download the snapshot block from).

Note that once the snapshot block for height `HEIGHT_SNAPSHOT+1` is processed, the node should start its regular block synchronization, i.e., fetching the blocks with height larger than `HEIGHT_SNAPSHOT+1`. The steps 3 to 5 from above could run in the background with low priority.

###### Minimal Validation of Core v3 Blocks

> A block created by Lisk Core v3, i.e., a block with a height between `HEIGHT_PREVIOUS_SNAPSHOT` and`HEIGHT_SNAPSHOT` (inclusive) is validated as follows:

Author: Typo fixed.

- Verify that the block follows the [`block schema`](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0029.md#deserialization) defined LIP 0029.
- Compute the block ID as defined in [LIP 0029](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0029.md#block-id) and verify that it is equal to the `previousBlockID` property of the child block.
- Verify that the transactions in the payload yield the [transaction root](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0032.md#specification) provided in the block header.

> If any of the steps fails, the block is invalid.

Reviewer 2: ```suggestion
If any of the above steps fail, the block is invalid.
```

Author: Fixed.

###### Assets Entry for Legacy Module

The `assets` entry for the legacy module is added by the logic defined in the function `addLegacyModuleEntry` in the following pseudo code:

```python
addLegacyModuleEntry():
    legacyObj = {}
    legacyObj.accounts = []
    for every account in legacyAccounts:
        userObj = {}
        userObj.address = account.address
        userObj.balance = account.token.balance
        legacyObj.accounts.append(userObj)

>     data = serialization of legacyObj using genesisLegacyStoreSchema

Reviewer 2: The reference URL to `genesisLegacyStoreSchema` is missing. Please consider adding them as a note below the code block.

Author: Added a reference to the schema.

    append {"moduleID": MODULE_ID_LEGACY, "data": data} to b.assets
```

###### Assets Entry for Token Module

The `assets` entry for the token module is added by the logic defined in the function `addTokenModuleEntry` in the following pseudo code:

```python
addTokenModuleEntry():
    tokenObj = {}
    tokenObj.userSubstore = createUserSubstoreArray()
    tokenObj.supplySubstore = createSupplySubstoreArray()
    tokenObj.escrowSubstore = []
    tokenObj.availableLocalIDSubstore = {}
    tokenObj.availableLocalIDSubstore.nextAvailableLocalID = LOCAL_ID_LSK + 1
    tokenObj.terminatedEscrowSubstore = []

>     data = serialization of tokenObj using genesisTokenStoreSchema

Author: Added a reference to the schema.

>     tokenObj.terminatedEscrowSubstore = []

Reviewer: I am not sure this substore still exists.

Author: Removed.

    append {"moduleID": MODULE_ID_TOKEN, "data": data} to b.assets

createUserSubstoreArray():
    userSubstore = []
    for every account in accounts:
        userObj = {}
        userObj.address = account.address
        userObj.tokenID = TOKEN_ID_LSK_MAINCHAIN
        userObj.availableBalance = account.token.balance
        userObj.lockedBalances = getLockedBalances(account)

>         if (userObj.availableBalance == 0 and userObj.lockedBalances == []):

Reviewer: As discussed, this was changed to also add zero balance accounts. Can be resolved.

            continue
        userSubstore.append(userObj)
    return userSubstore

getLockedBalances(account):
    amount = 0
    for vote in account.dpos.sentVotes:
        amount += vote.amount
    for unlockingObj in account.dpos.unlocking:
        amount += unlockingObj.amount
    if amount > 0:
        return [{"moduleID": MODULE_ID_DPOS, "amount": amount}]
    else:
        return []

createSupplySubstoreArray():
    totalLSKSupply = 0
    for every account in accounts:

>     totalLSKSupply += account.token.balance

Reviewer: The indentation is incorrect here.

Author: Fixed!

    lockedBalances = getLockedBalances(account)
    if lockedBalances is not empty:
        totalLSKSupply += lockedBalances[0].amount
    LSKSupply = {"localID": LOCAL_ID_LSK, "totalSupply": totalLSKSupply}
    return [LSKSupply]
```

###### Assets Entry for Auth Module

The `assets` entry for the auth module is added by the logic defined in the function `addAuthModuleEntry` in the following pseudo code:

```python
addAuthModuleEntry():
    authDataSubstore = []
    for every account in accounts:
        authObj = {}
        authObj.numberOfSignatures = account.keys.numberOfSignatures
        authObj.mandatoryKeys = account.keys.mandatoryKeys
        authObj.optionalKeys = account.keys.optionalKeys
        authObj.nonce = account.sequence.nonce
        entry = {address: account.address, authAccount: authObj}
        authDataSubstore.append(entry)

>     data = serialization of authDataSubstore using genesisAuthStoreSchema

Reviewer 2: Same for `genesisAuthStoreSchema`.

Author: Added a reference to the schema.

    append {"moduleID": MODULE_ID_AUTH, "data": data} to b.assets
```

###### Assets Entry for DPoS Module

The `assets` entry for the DPoS module is added by the logic defined in the function `addDPoSModuleEntry` in the following pseudo code:

```python
addDPoSModuleEntry():
    DPoSObj = {}
    DPoSObj.validators = createValidatorsArray()
    DPoSObj.voters = createVotersArray()
    DPoSObj.snapshots = []
    DPoSObj.genesisData = createGenesisDataObj()

>     data = serialization of DPoSObj using genesisDPoSStoreSchema

Author: Added a reference to the schema.

    append {"moduleID": MODULE_ID_DPOS, "data": data} to b.assets

createValidatorsArray():
    validators = []
    validatorKeys = getValidatorKeys()
    for every account in accounts:
        if account.dpos.delegate.username == "":
            continue
        validator = {}
        validator.address = account.address
        validator.name = account.dpos.delegate.username

>         validator.blsKey = INVALID_BLS_KEY

Author: Added a reference.

>         validator.blsKey = INVALID_BLS_KEY

Reviewer: I would also add it to the table as the beginning for consistency as we did the same for other constants.

>         validator.blsKey = INVALID_BLS_KEY

Reviewer: Looks good. I resolve the comment.

    return validators

# This functions gets the public keys of the registered validators,
# i.e., the accounts for which account.dpos.delegate.username is not
# the empty string, from the history of Lisk Core v3 blocks and
# transactions.
getValidatorKeys():
    let keys be an empty dictionary
    for every block c with height in [HEIGHT_PREVIOUS_SNAPSHOT_BLOCK + 1,  HEIGHT_SNAPSHOT]:
        address = address(c.generatorPublicKey)
        keys[address] = c.generatorPublicKey
        for trs in c.transactions:
            address = address(trs.senderPublicKey)
            if the validator corresponding to address is a registered validator:
                keys[address] = trs.senderPublicKey
    return keys

createVotersArray():
    voters = []
    for every account in accounts:
        if account.dpos.sentVotes == [] and account.dpos.unlocking == []:
            continue
        voter = {}
        voter.address = account.address
        voter.sentVotes = account.dpos.sentVotes
        voter.pendingUnlocks = account.dpos.unlocking
        voters.append(voter)
    return voters

createGenesisDataObj():
    genesisDataObj = {}
    genesisDataObj.initRounds = DPOS_INIT_ROUNDS

>     r = round number for the block a

Reviewer: It should be possible to deterministically compute r from the height of block a. I would add this to the specifications here.

Author: Added for clarity.

Reviewer: This is how we compute the round number in LIP 57. Did you check with someone from the SDK team if that is also the same currently in Lisk Core 3.0?

>     Let topDelegates be the set of the top 101 non-banned delegate accounts by delegate weight at the end of round r-2

Reviewer: ```suggestion
    let topDelegates be the set of the top 101 non-banned delegate accounts by delegate weight at the end of round r-2
```

Author: Fixed!

>     initDelegates = [account.address for account in topDelegates]

Reviewer: We should uniquely define the order in this array.

    sort initDelegates in lexicographical order
genesisDataObj.initDelegates = initDelegates

> return genesisDataObj

Reviewer: The indentation of the two lines above is incorrect.

Author: Fixed!

To avoid the need to send an _updated generator key_ right after the hard fork, registered validators that did not forge a block and did not send a transaction in the Lisk Core v3 history could simply send a transaction any time before the hard fork. Then the sender public key of this transaction will be set as the generator key as described above. Nevertheless, every validator is encouraged to perform an _update generator key_ with a generator key different from the public key used for transaction signatures eventually. This is due to [security benefits](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) added by the concept of generator keys.

> #### Invalid Ed25519 Public Key

Reviewer: @sergeyshemyakov Did you verify this part also for the library that we use in our implementation? If not, I think it would be good if you could do that.

Author: I have checked, and the public key should be different. The little endian encoding of 2^255-1 should be `0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffe`, note that little endiannes holds bitwise as described in the end of this section: https://datatracker.ietf.org/doc/html/rfc8032#section-2.
I have asked Shu about the implementation library, I will also try to check the actual behavior of SDK.

   </td>
   <td>TBD
   </td>
   <td>Module ID of the <a href="https://github.com/LiskHQ/lips/blob/main/proposals/lip-0041.md">Auth Module</a>
   </td>
  </tr>
  <tr>
   <td><code>MODULE_ID_TOKEN</code>
   </td>
   <td>bytes
   </td>
   <td>TBD
   </td>
   <td>Module ID of the <a href="https://github.com/LiskHQ/lips/blob/main/proposals/lip-0051.md">Token Module</a>
   </td>
  </tr>
  <tr>
   <td><code>MODULE_ID_LEGACY</code>
   </td>
   <td>bytes
   </td>
   <td>TBD
   </td>
   <td>Module ID of the <a href="https://github.com/LiskHQ/lips/blob/main/proposals/lip-0050.md">Legacy Module</a>.
   </td>
  </tr>
  <tr>
   <td><code>INVALID_ED25519_KEY</code>
   </td>
   <td>bytes
   </td>
   <td>

```   
0xffffffff
ffffffffff
ffffffffff
ffffffffff
ffffffffff
ffffffffff
ffffff
```

   </td>
   <td>An Ed25519 public key for which the signature validation always fails. This value is used for the <code>generatorPublicKey</code> property of validators in the snapshot block for which no public key is contained within the history since the last snapshot block.
   </td>
  </tr>
  <tr>
   <td><code>INVALID_BLS_KEY</code>
   </td>
   <td>bytes
   </td>
   <td>

48 bytes all set to 0

   </td>

>    <td>An invalid BLS key, used as a placeholder before a valid BLS key is registered. It is invalid as deserialization fails since the most significant bit of the first byte is zero while the total length is 48 (see second point <a href="https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-07#appendix-C.2">here</a>).

Author: Agree.

[Fee Module](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0048.md#notation-and-constants)             | `TOKEN_ID_FEE`                 | `TOKEN_ID_LSK = 0x 00 00 00 00 00 00 00 00`

### Mainnet Migration

#### Process

![Overview of migration process](lip_mainnet_migration/migration_graphic.png)

_Figure 1: Overview of the migration process. Elements in blue are created by Lisk Core v3, including the state snapshot. Elements in yellow - the snapshot block - are created by the migrator tool. Elements in green are created by Lisk Core v4._

The migration from Lisk Core v3 to Lisk Core v4, also depicted in Figure 1, is performed by the following steps:

1. Nodes run Lisk Core v3, where the following steps are done:
    1. After a block at height `HEIGHT_SNAPSHOT` was processed, a snapshot of the state is performed, which we denote by `STATE_SNAPSHOT`. If this block is reverted and a new block for this height is processed, then also `STATE_SNAPSHOT` needs to be computed again.
    2. Nodes continue processing blocks until the block at height `HEIGHT_SNAPSHOT` is final.

>     3. Once the block at height `HEIGHT_SNAPSHOT` is final, nodes can stop forging and processing new blocks. Finalized blocks with a height larger than or equal to `HEIGHT_SNAPSHOT + 1` are discarded.

Author: Good point, all blocks with a height larger or equal than `HEIGHT_SNAPSHOT + 1` are discarded, independent of the finality status. Will phrase this better!

Note that once the snapshot block for height `HEIGHT_SNAPSHOT + 1` is processed, the node should start its regular block synchronization, i.e., fetching the blocks with height larger than `HEIGHT_SNAPSHOT + 1`. The steps 3 to 5 from above could run in the background with low priority.

###### Minimal Validation of Core v3 Blocks

A block created by Lisk Core v3, i.e., a block with a height between `HEIGHT_PREVIOUS_SNAPSHOT` and `HEIGHT_SNAPSHOT` (inclusive) is validated as follows:

- Verify that the block follows the [`block schema`](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0029.md#deserialization) defined LIP 0029.
- Compute the block ID as defined in [LIP 0029](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0029.md#block-id) and verify that it is equal to the `previousBlockID` property of the child block.
- Verify that the transactions in the payload yield the [transaction root](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0032.md#specification) provided in the block header.

> If any of the steps above fails, the block is invalid.

Reviewer: ```suggestion
If any of the steps above fail, the block is invalid.
```

    r = ceiling((a.header.height - HEIGHT_PREVIOUS_SNAPSHOT_BLOCK) / ROUND_LENGTH)
    let topDelegates be the set of the top 101 non-banned delegate accounts by delegate weight at the end of round r-2
    initDelegates = [account.address for account in topDelegates]
    sort initDelegates in lexicographical order
    genesisDataObj.initDelegates = initDelegates
    return genesisDataObj
```

## Rationale

### Keeping Lisk Core v3 Block History

The decision to discard the block history of Lisk Mainnet at the hard fork from Lisk Core v2 to Lisk Core v3 was considered as rather disadvantageous. One reason is that the transaction history of an account can not be retrieved anymore from a node running Lisk Core v3\. Therefore, the block history of blocks created with Lisk Core v3 shall be kept on Lisk Core v4 nodes.

#### Getting the Lisk Core v3 Block History

In order to keep the implementation of Lisk Core v4 as clean as possible, Lisk Core v4 cannot process Lisk Core v3 blocks, but only ensure their integrity. This is, however, sufficient for maintaining the Lisk Core v3 block history and guaranteeing the immutability of this block history.

> Once Lisk Core v4 processed the snapshot block, it can fetch the whole Lisk Core v3 block history, where this is done from the highest to the lowest height. This allows that each block can be validated by simply checking if its block ID matches with the `previousBlockID` property of the child block, and if the payload matches with the `transactionRoot` property. Thus, almost no protocol rules related to Lisk Core v3 must be implemented which keeps the implementation clean. The Lisk Core v3 history can be fetched from a database created by Lisk Core v3 on the same computer or from the peer-to-peer network. Note that the first Lisk Core v3 block - the snapshot block of the previous migration - poses an exception as this one is either [downloaded from a server](#fetching-snapshot-blocks) or fetched from a database created by Lisk Core v3 on the same computer.

Reviewer 3: ```suggestion
Once Lisk Core v4 processes the snapshot block, it can fetch the whole Lisk Core v3 block history, where this is done from the highest to the lowest height. This allows that each block can be validated by simply checking if its block ID matches with the `previousBlockID` property of the child block, and if the payload matches with the `transactionRoot` property. Thus, almost no protocol rules related to Lisk Core v3 must be implemented which keeps the implementation clean. The Lisk Core v3 history can be fetched from a database created by Lisk Core v3 on the same computer or from the peer-to-peer network. Note that the first Lisk Core v3 block - the snapshot block of the previous migration - poses an exception as this one is either [downloaded from a server](#fetching-snapshot-blocks) or fetched from a database created by Lisk Core v3 on the same computer.
```
Maybe better to use the term "machine" instead of "computer" since microcontroller, ASIC  etc. may not fall under the same roof but can be used nonetheless.

Author: Fixed, also changed "computer" to "machine".

### Fetching Snapshot Blocks

> The snapshot block created for this hard fork as well as for the [one used for the previous hard fork](https://lisk.observer/block/abdea6dfe5941922d958d1f1dc1d2f97249723b8852325334e55b73e92584191) are not shared via peer-to-peer due to their large size. Instead, they are either downloaded from a server or must be found locally.

Reviewer 3: ```suggestion
The snapshot block created for this hard fork as well as for the [one used for the previous hard fork](https://lisk.observer/block/abdea6dfe5941922d958d1f1dc1d2f97249723b8852325334e55b73e92584191) is not shared via peer-to-peer network due to their large size. Instead, they are either downloaded from a server or must be found locally.
```

#### Initial Version of Lisk Core v4

> The initial version of Lisk Core v4 can not download the snapshot block of this hard fork from a server because it could not validate the block. Instead, it must find the snapshot block locally on the same computer. This requires that the snapshot block must be created using a migrator tool on the same node, which in turn requires that the node runs the latest patched version of Lisk Core v3 until the block at height `HEIGHT_SNAPSHOT` is final.

Reviewer 3: ```suggestion
The initial version of Lisk Core v4 can not download the snapshot block for this hard fork from a server because it could not validate it. Instead, it must find the snapshot block locally on the same computer. This requires that the snapshot block must be created using a migrator tool on the same node, which in turn requires that the node runs the latest patched version of Lisk Core v3 until the block at height `HEIGHT_SNAPSHOT` is final.
```
also please consider "computer" vs "machine"

Author: Changed "computer" to "machine" everywhere.

#### Later Versions of Lisk Core v4

> Once the snapshot block is final, a patched version of Lisk Core v4 will be released in which the block ID of the snapshot block is hard coded. This allows to validate a downloaded snapshot block by simply checking its block ID.

Reviewer 3: ```suggestion
Once the snapshot block is final, a patched version of Lisk Core v4 can be released in which the block ID of the snapshot block is hard coded. This allows to validate a downloaded snapshot block by simply checking its block ID.
```
to align with the [previous](https://github.com/LiskHQ/lips/blob/lip-mainnet-migration/proposals/mainnet_migration.md#fetching-block-history-and-snapshot-block-without-running-lisk-core-v3) statement or vice-versa

#### Downloading from Server

> The URL can be configured. By default, the URL points to a server of the Lisk Foundation. The snapshot block of this hard fork is validated by checking that its block ID matches the hard coded block ID in the patched version of Lisk Core v4\. The snapshot block of the previous migration is validated by checking that its block ID matches with the `previousBlockID` property of its child block (see also [here](#getting-the-lisk-core-3-block-history)). Hence, there are no security concerns by downloading them from a non-trusted server. Users can nevertheless configure their node to download it from a server of their choice. The snapshot blocks could be provided by any users. This may also be helpful for situations where the Lisk Foundation server should be down.

Reviewer 3: ```suggestion
The URL can be configured. By default, the URL points to a server of the Lisk Foundation. The snapshot block for this hard fork is validated by checking that its block ID matches the hard coded block ID in the patched version of Lisk Core v4\. The snapshot block of the previous migration is validated by checking that its block ID matches with the `previousBlockID` property of its child block (see also [here](#getting-the-lisk-core-3-block-history)). Hence, there are no security concerns by downloading them from even a non-trusted server. Users can nevertheless configure their node to download it from a server of their choice. The snapshot blocks could be provided by any user. This may also be helpful in situations where the Lisk Foundation server is down.
```

### Assigning Generator Key to Validator Accounts

> In the protocol used by Lisk Core v4, each validator needs a [generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) in order to generate blocks. Moreover, the DPoS module expects that each validator account has such a key specified in the snapshot block. Otherwise, the snapshot block is invalid (see [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-state-initialization)). Therefore, it is aimed to use the existing public keys of the validator accounts as the initial generator keys in the snapshot block. However, the public key does not belong to the [account state](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html) anymore with Lisk Core v3\. For this reason, the whole block history from the previous snapshot block onwards is scanned for public keys of registered validators. The found keys are used as generator keys in the snapshot block. Validator accounts for which no public key was found get an "invalid" public key assigned, i.e., one for which every signature validation will fail. The specific value is discussed in the [following subsection](#invalid-ed25519-public-key). Those validators should send an [update generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#update-generator-key) transaction as soon as possible once Lisk Core v4 is used, ideally within the [bootstrap period](#bootstrap-period). Otherwise, those validators can still be active validators, but are not able to generate valid blocks as they cannot create [valid block signatures](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md#signature). In the worst case, this may lead to banning which happens after 50 blocks in a row are missed.

Reviewer 3: `However, the public key does not belong to the account state anymore with Lisk Core v3`
not sure what this implies - did it belong to the account state with v3 in the past ?

Author: As I understand [this documentation](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html), the public key used to be part of account.

However, I don't know the history of the protocol so well.

> In the protocol used by Lisk Core v4, each validator needs a [generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) in order to generate blocks. Moreover, the DPoS module expects that each validator account has such a key specified in the snapshot block. Otherwise, the snapshot block is invalid (see [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-state-initialization)). Therefore, it is aimed to use the existing public keys of the validator accounts as the initial generator keys in the snapshot block. However, the public key does not belong to the [account state](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html) anymore with Lisk Core v3\. For this reason, the whole block history from the previous snapshot block onwards is scanned for public keys of registered validators. The found keys are used as generator keys in the snapshot block. Validator accounts for which no public key was found get an "invalid" public key assigned, i.e., one for which every signature validation will fail. The specific value is discussed in the [following subsection](#invalid-ed25519-public-key). Those validators should send an [update generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#update-generator-key) transaction as soon as possible once Lisk Core v4 is used, ideally within the [bootstrap period](#bootstrap-period). Otherwise, those validators can still be active validators, but are not able to generate valid blocks as they cannot create [valid block signatures](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md#signature). In the worst case, this may lead to banning which happens after 50 blocks in a row are missed.

Reviewer 3: ```suggestion
In the protocol used by Lisk Core v4, each validator needs a [generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0044.md#generator-key) in order to generate blocks. Moreover, the DPoS module expects that each validator account has such a key specified in the snapshot block. Otherwise, the snapshot block is invalid (see [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-state-initialization)). Therefore, it is aimed to use the existing public keys of the validator accounts as the initial generator keys in the snapshot block. However, the public key does not belong to the [account state](https://lisk.com/documentation/lisk-sdk/protocol/accounts.html) anymore with Lisk Core v3\. For this reason, the whole block history from the previous snapshot block onwards is scanned for public keys of registered validators. The found keys are used as generator keys in the snapshot block. Validator accounts for which no public key was found are assigned an "invalid" public key, i.e., one for which every signature validation will fail. The specific value is discussed in the [following subsection](#invalid-ed25519-public-key). These validators should send an [update generator key](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#update-generator-key) transaction as soon as possible once Lisk Core v4 is used, ideally within the [bootstrap period](#bootstrap-period). Otherwise, they can still be active validators, but not able to generate valid blocks as they cannot create [valid block signatures](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0055.md#signature). In the worst case, this may lead to banning which happens after 50 blocks are missed in a row.
```

### Bootstrap Period

> There will be a bootstrap period in which a fixed set of delegates will constitute the set of active delegates. This period will last for one week assuming no missed blocks. This is done by setting the [`initRounds`](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-assets-schema) property in the snapshot block to the corresponding value, which is given by `DPOS_INIT_ROUNDS`. The reason to have this period is that the DPoS module selects only delegates with a valid BLS key outside of the bootstrap period (there will be an update to [LIP 0057](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md) that adapts the delegate selection in this way). Thus, without bootstrap period, the DPoS module could not select any delegates right after the hard fork which would result in a stop of the chain.

Reviewer 3: ```suggestion
There will be a bootstrap period in which a fixed set of delegates will constitute the set of active delegates. This period will last for one week assuming no missed blocks. This is done by setting the [`initRounds`](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#genesis-assets-schema) property in the snapshot block to the corresponding value, which is given by `DPOS_INIT_ROUNDS`. The reason behind having this period is that the DPoS module selects delegates only with a valid BLS key outside of the bootstrap period (there will be an update to [LIP 0057](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md) that adapts the delegate selection in this way). Thus, without a bootstrap period, the DPoS module could not select any delegates right after the hard fork which would result in a stop of the chain.
```
